Add loading screen/circle

timhitchins opened the pull request:
This starts to address part 2 of issue #61: 

>Add a loading indicator somewhere to the map to give the user feedback that the map is loading. This would be especially helpful for folks on slower network connections.

Introduces the following:
- new loading component (I left comments for now)
- additional scss for loading circle
- additional html in index.html

The loading indicator will become hidden after all the layers have resolved. The component can be built out further to become hidden after specific layers have loaded, but for now will only be hidden after all layers have loaded.  This is because the layer controller/toggle box doesn't become visible until after all the layers have loaded.  I am also not tied to the loading circle itself if others have a better option in mind.

Edit :
- I simplified the component by removing map layer logic to make it more flexible as well as removed dead code.
- I updated the styles for centering the indicator
- I added a indicator toggle in the map
When/if we decide to lazy load the layers, this component will be much more flexible.

clhenrick commented on src/components/Loading.js:
```diff
@@ -0,0 +1,42 @@
+  bindListeners() {
+    //layer can be passed in here to keep track of
+    //which layers have loaded
+    dispatch.on("fetch-map-data-resolve", (/*layer*/) => {
```
It would probably make sense to name space this event, e.g. `"fetch-map-data-resolve.loading"` as not name spacing it will prevent the event from not being registered elsewhere if another listener also uses `"fetch-map-data-resolve"` without a name space. See the readme for d3-dispatch for more.

timhitchins replied:
Definitely all valid comments especially on the re-usability.  Will work on this one.

clhenrick commented on src/components/Loading.js:
```diff
@@ -0,0 +1,42 @@
+  // loaded layers counter
+  layersLoaded = 0;
+
+  // optionally pass in the mapconfig to keep
```
Currently the way in which new layers are added to the map happens in such a way that if a layer is set to `false` in the `mapConfig` a `fetch` request will still be made for that layer. This means that the map will still be waiting for all layers to finish fetching, because even if they're not added to the map they are still cached in the map's `dataLayers`. It might be better to only `fetch` a layer's data if it is enabled when the map loads, and then when a layer is toggled on for the first time, fetch the data. 

Anyway, I bring this up as I'm not sure if that's what you're already thinking about / planning for. It's my preference to not leave dead code in for future work as to not confuse others.

timhitchins replied:
That about what I was thinking.  Yeah I noticed.  With changes mentioned from above this will be addressed.

clhenrick commented on src/components/Loading.js:
```diff
@@ -0,0 +1,42 @@
+  TOTAL_NUMBER_OF_MAP_LAYERS,
+} from "utils/constants";
+
+export class Loading {
```
maybe a better name might be `LoadingIndicator` to help disambiguate?

timhitchins replied:
Definitely good point.

clhenrick commented on src/styles/_loading.scss:
```diff
@@ -0,0 +1,41 @@
+  animation: spin 1s linear infinite;
+}
+
+@-webkit-keyframes spin {
```
I checked the CSS animation on IE11 & Firefox, and it looks good. Haven't yet checked mobile, Safari, or Chrome.

timhitchins replied:
Nice.

timhitchins commented:
Some updates:
- I simplified the component by removing map layer logic to make it more flexible as well as removed dead code. 
- I updated the styles for centering the indicator
- I added a indicator toggle in the map

When/if we decide to lazy load the layers, this component will be much more flexible.

clhenrick commented on src/components/LeafletMap.js:
```diff
@@ -120,6 +120,7 @@ export class LeafletMap {
 
   handleAddLayerError = (error) => {
     console.error(error);
+    this.toggleLoadingIndicator(true); //show loading indicator
```
@timhitchins wouldn't we want to do the opposite here? If a layer fails to fetch we don't want the user to be stuck with a modal overlay that's blocking them from interacting with the map.

timhitchins replied:
Good point.

clhenrick commented on src/components/LeafletMap.js:
```diff
@@ -257,4 +261,12 @@ export class LeafletMap {
     });
   };
+
+  toggleLoadingIndicator = (isLoading) => {
```
@timhitchins this could just be an `if/else` rather than short circuit with a `return`, IMO the logic is more clear that way, but just a suggestion.

timhitchins replied:
bad habit.  good point.

clhenrick commented on src/components/LoadingIndicator.js:
```diff
@@ -0,0 +1,21 @@
+
+  show() {
+    // the default for the corresponding DOM element is false
+    document.getElementById("loading-container").setAttribute("hidden", false);
```
@timhitchins you could add the HTML node as a class property to be a little more DRY here, e.g.

```js
export class LoadingIndidcator {
  node = document.getElementById("loading-container");
// ...
```

timhitchins replied:
dry is def better. good one.

timhitchins commented:
@clhenrick whenever you get a min can you approve so merging isn't blocked.

timhitchins commented:
Thanks @jimboweb.  I imagine it will be easier to work with once you get going with all the layer fetching.
